# set_exception_handler




<div class="phpcode"><span class="html">
Things you should be aware of:<br><br>An exception handler handles exceptions that were not caught before. It is the nature of an exception that it discontinues execution of your program - since it declares an exceptional situation in which the program cannot continue (except you catch it).<br><br>Since it has not been catched your code signals it is not being aware of the situation and cant go on.<br><br>This implies: returning to the script is simply impossible when the exception handler has already been called, since an uncaught exception is not a notice. use your own debug- or notice-log-system for things like that.<br><br>Furthermore: While is is still possible to call functions from your script, since the exception handler has already been called exceptions bubbling from that piece of code won&apos;t trigger the exception handler again. php will die without leaving any information apart form &quot;uncaught exception with unknown stack frame&quot;. So if you call functions from your script, make sure that you catch any exceptions that possibly occur via try..catch inside the exception handler.<br><br>For those of you who misinterpreted the essential meaning of the exception handler: it&apos;s only use is to handle the abortion of your script gracefully, e.g. in a project like facebook or wikipedia: render a nice error page, eventually hiding information which shall not leak into the public (instead you may want to write to your log or mail the sys-admin or stuff like that).<br><br>In other words: Redirecting all php-errors form an error-handler using exceptions - including notices - is a very dumb idea, if you do not intend having your script aborted everytime you didn&apos;t set a variable (for example).<br><br>my 2 cents.</span>
</div>
  

#


<div class="phpcode"><span class="html">
If you want a class instance to handle the exception, this is how you do it :<br><br><span class="default">&lt;?php<br></span><span class="keyword">class </span><span class="default">example </span><span class="keyword">{<br>&#xA0;&#xA0; public function </span><span class="default">__construct</span><span class="keyword">() {<br>&#xA0; &#xA0; &#xA0;&#xA0; @</span><span class="default">set_exception_handler</span><span class="keyword">(array(</span><span class="default">$this</span><span class="keyword">, </span><span class="string">&apos;exception_handler&apos;</span><span class="keyword">));<br>&#xA0; &#xA0; &#xA0;&#xA0; throw new </span><span class="default">Exception</span><span class="keyword">(</span><span class="string">&apos;DOH!!&apos;</span><span class="keyword">);<br>&#xA0;&#xA0; }<br><br>&#xA0;&#xA0; public function </span><span class="default">exception_handler</span><span class="keyword">(</span><span class="default">$exception</span><span class="keyword">) {<br>&#xA0; &#xA0; &#xA0;&#xA0; print </span><span class="string">&quot;Exception Caught: &quot;</span><span class="keyword">. </span><span class="default">$exception</span><span class="keyword">-&gt;</span><span class="default">getMessage</span><span class="keyword">() .</span><span class="string">&quot;\n&quot;</span><span class="keyword">;<br>&#xA0;&#xA0; }<br>}<br><br></span><span class="default">$example </span><span class="keyword">= new </span><span class="default">example</span><span class="keyword">;<br><br></span><span class="default">?&gt;<br></span><br>See the first post (Sean&apos;s) for a static example.&#xA0; As Sean points out, the exception_handler function must be declared public.</span>
</div>
  

#


<div class="phpcode"><span class="html">
A behaviour not documented or discussed enough, yet pretty common is that is that if an exception is thrown from the global exception handler then a fatal error occurs (Exception thrown without a stack frame). That is, if you define your own global exception handler by calling set_exception_handler() and you throw an exception from inside it then this fatal error occurs. It is only natural though, as the callback defined by set_exception_handler() is only called on uncaught (unhandled) exceptions so if you throw one from there then you get this fatal error as there is no exception handler left (you override the php internal one by calling set_exception_handler()), hence no stack frame for it.<br><br>Example:<br><br><span class="default">&lt;?php<br><br></span><span class="keyword">function </span><span class="default">myExceptionHandler </span><span class="keyword">(</span><span class="default">Exception $ex</span><span class="keyword">)<br>{<br>&#xA0; &#xA0; throw </span><span class="default">$ex</span><span class="keyword">;<br>}<br><br></span><span class="default">set_exception_handler</span><span class="keyword">(</span><span class="string">&quot;myExceptionHandler&quot;</span><span class="keyword">);<br><br>throw new </span><span class="default">Exception</span><span class="keyword">(</span><span class="string">&quot;This should cause a fatal error and this message will be lost&quot;</span><span class="keyword">);<br><br></span><span class="default">?&gt;<br></span><br>Will cause a Fatal error: Exception thrown without a stack frame<br><br>If you skip/comment the set_exception_handler(&quot;...&quot;) line then the internal PHP global handler will catch the exception and output the exception message and trace (as string) to the browser, allowing you to at least see the exception message.<br><br>While it is a very good idea to always define your own global exception handler by using the set_exception_handler() function, you should pay attention and never throw an exception from it (or if you do then catch it).<br><br>Finally, every serious coder should use an IDE with debugging capabilities. Tracking down an error like this becomes a trivial matter by using simple debugging &quot;Step into&quot; commands (I for one recommend Zend IDE v5.2 at the moment of this writing). I have seen numerous messages on the internet with people wondering why this message pops up.<br><br>Cheers<br><br>p.s. Other causes for this error which are somehow unrelated to this is when you throw an exception from a destructor (the reasons behind that are similar though, the global handler might no longer exist due to the php engine shutting the page down).</span>
</div>
  

#

[Official documentation page](https://www.php.net/manual/en/function.set-exception-handler.php)

**[â¬† to root](/)**